src/infinitempomatrix.jl

```diff
@@ -43,4 +44,91 @@ function InfiniteMPOMatrix(data::Vector{Matrix{ITensor}}, translator::Function)
   return InfiniteMPOMatrix(CelledVector(data, translator), 0, size(data)[1], false)
 end
 
-#nrange(H::InfiniteMPOMatrix) = [size(H[j])[1] - 1 for j in 1:nsites(H)]
+#
+#  Hm should have the form below.  Only link indices are shown.
+#
+#    I         0                     0           0      0
+#  M-->l=n     0                     0           0      0
+#    0    l=n-->M-->l=n-1 ...        0           0      0
+#    :         :                     :           :      :
+#    0         0          ...  l=2-->M-->l=1     0      0
+#    0         0          ...        0        l=1-->M   I
+#
+#  We need to capture the corner I's and the sub-diagonal Ms, and paste them together into an ITensor.
+#  This is facilutated by making all elements of Hm into order(4) tensors by adding dummy Dw=1 indices.
+#  We ITensors.directsum() to join all the blocks into the final ITensor.
+#  This code should be 100% dense/blocks-sparse agnostic.
+#
+function matrixITensorToITensor(Hm::Matrix{ITensor})::ITensor
+  indexT = typeof(inds(Hm[1, 1])[1])
+  T = eltype(Hm[1, 1])
+  lx, ly = size(Hm)
+  @assert lx == ly
+  #
+  #  Extract the sub diagonal
+  #
+  Ms = map(n -> Hm[lx - n + 1, ly - n], 1:(lx - 1)) #Get the sub diagonal into an array.
+  N = length(Ms)
+  @assert N == lx - 1
+  #
+  #  We need a dummy link index, but it has to have the right direction.  
+  #  Ms[1] should only have one link index which should be pointing to the right.
+  #
+  @assert length(inds(Ms[1]; tags="Link")) == 1
```

Comment:
Please avoid using `tags` in this way for filtering or extracting indices (that goes for the rest of the PR as well).

Reply:
Assuming you are talking about the semicolon: Done

Reply:
No, I meant a deeper change, that instead of finding a certain Index using the tags of the Index, we should find indices by searching which indices of the tensor are shared with neighboring tensors in the network. So for example in a line like this:
```julia
il, = inds(Ms[n], tags=tags(ir)) #Find the left index
```
we should have instead already determined which are the left and right indices of the MPO tensors by analyzing the MPO networks and extracting the `linkinds` and then passing those indices into this function (maybe there is a unique way to determine which indices are getting direct summed over/are the link indices instead of the site indices without passing indices into this function and without using the tags of the indices, but I doubt it).

Basically, we should take as an axiom or design "law" of ITensor code that we should not use tags to find indices, filter indices, decide which indices to prime, etc. and instead use tensor network analysis tools or index set operation like `linkinds`, `commoninds`, `uniqueinds`, etc. (our versions of what Julia would call Set operations: https://docs.julialang.org/en/v1/base/collections/index.html#Set-Like-Collections). This basically means not using the `tags` keyword argument for filtering indices _at all_. I have found that designing code in this way is much more robust, leads to simpler code in the end, the code is less error prone, etc.

Tags should just be treated as being used mostly for printing, or as ways to avoid having indices contract with each other (which is the purpose of the call tags of the InfiniteMPS).

Reply:
For the record, for passing keyword arguments to functions, I prefer the semicolon syntax. This is also consistent with the "Blue" style guide that we follow in our code and which is enforced by the auto formatter: https://github.com/invenia/BlueStyle#keyword-arguments

Reply:
I think this would require mods to the generation of InfiniteMPOMatrix:
Hm = InfiniteMPOMatrix(model, s; model_params...);
Ms = map(n -> Hm[1][lx - n + 1, ly - n], 1:(lx - 1))
@show inds(Ms[1],tags="Link")
@show inds(Ms[2],tags="Link")
@show inds(Ms[3],tags="Link")

inds(Ms[1], tags = "Link") = ((dim=5|id=553|"Link,l=1") 
inds(Ms[2], tags = "Link") = ((dim=5|id=154|"Link,l=1")  (dim=7|id=906|"Link,l=2") 
inds(Ms[3], tags = "Link") = ((dim=7|id=582|"Link,l=2")  (dim=11|id=172|"Link,l=3") 

The IDs don't line up, only the tags.

Reply:
The `InfiniteMPOMatrix` was constructed from sets of MPOs which had common indices, so there must be common indices across the sites of the `InfiniteMPOMatrix`. Specifically, if you did a matrix multiplication of the `Matrix{ITensor}`s on site `1` and `2`, i.e. `Hm[1] * Hm[2]`, shouldn't the tensors making up the matrices `Hm[1]` and `Hm[2]` share common link indices? Those are the common indices I'm taking about. I don't think we need to (or should) assume that within a certain `Matrix{ITensor}`, say `Hm[1]`, that the tensors share common link indices (though they should share common site indices). We should pass information into the function about which indices are shared between `Hm[1]` and `Hm[2]`, for example, and then direct sum over those indices.

Reply:
OK I get it now ... thanks for your patience.

Reply:
Done